argmin-math/src/lib.rs

```diff
@@ -275,3 +279,102 @@ pub trait ArgminMinMax {
     /// Select piecewise maximum
     fn max(x: &Self, y: &Self) -> Self;
 }
+
+/// Defines associated types
+pub trait ArgminTransition {
+    /// Associated type for 1-D array
+    type Array1D;
+    /// Associated type for 2-D array
+    type Array2D;
+}
+
+/// Create a random array
+pub trait ArgminRandomMatrix {
+    /// Draw samples from a standard Normal distribution (mean=0, stdev=1)
+    fn standard_normal(nrows: usize, ncols: usize) -> Self;
+}
+
+/// Get dimensions of an array
+pub trait ArgminSize<U> {
+    /// Get array shape
+    fn shape(&self) -> U;
+}
+
+/// Eigendecomposition
+pub trait ArgminEigen<U> {
+    /// Get the eigenvalues and eigenvectors of a complex Hermitian (conjugate symmetric) or a real symmetric matrix.
+    /// Returns a 1-D array containing the eigenvalues and a 2-D square array with eigenvectors (in columns).
+    fn eig_sym(&self) -> (U, Self);
+}
+
+/// Returns the indices that would sort an array.
+pub trait ArgminArgsort {
+    /// Perform sort and return an array of indices of the same shape.
+    fn argsort(&self) -> Vec<usize>;
+}
+
+/// Take elements from an array along an axis.
+pub trait ArgminTake<I> {
+    /// Take elements from indices along a given axis.
+    /// Returns as new array of the same size as indices
+    fn take(&self, indices: &[I], axis: u8) -> Self;
+}
+
+/// Defines iterator
+pub trait ArgminIter<T> {
```

Comment:
From the implementation it seems to me that instead of `ArgminIter` (and `ArgminMutIter`), one could maybe use `Iterator` directly. However, I do feel that this way it is more consistent with `ArgminAxisIter`, which is more complicated and hence makes sense to abstract over for different backends. I guess I'm just thinking out lout because I'm not sure what I want to say with this comment. I don't think that this should be changed.

Reply:
I don't think it will be safe to assume that every single type of 1-D array will implement `Iterator` trait. `vec` doesn't.

Reply:
Agree!